dev/set_matrix.py

```diff
@@ -0,0 +1,427 @@
+
+    return changed_flavors
+
+
+def divider(title, length=None):
+    r"""
+    Examples
+    --------
+    >>> divider("1234", 20)
+    '\n======= 1234 ======='
+    """
+    length = shutil.get_terminal_size(fallback=(80, 24))[0] if length is None else length
+    rest = length - len(title) - 2
+    left = rest // 2 if rest % 2 else (rest + 1) // 2
+    return "\n{} {} {}".format("=" * left, title, "=" * (rest - left))
+
+
+def str_to_operator(s):
+    """
+    Turns a string into the corresponding operator.
+
+    Examples
+    --------
+    >>> str_to_operator("<")(1, 2)  # equivalent to '1 < 2'
+    True
+    """
+    return {
+        # https://docs.python.org/3/library/operator.html#mapping-operators-to-functions
+        "<": operator.lt,
+        "<=": operator.le,
+        "==": operator.eq,
+        "!=": operator.ne,
+        ">=": operator.ge,
+        ">": operator.gt,
+    }[s]
+
+
+def get_operator_and_version(ver_spec):
+    """
+    Converts a version specifier (e.g. "< 3") to a tuple of (operator, version).
+
+    Examples
+    --------
+    >>> get_operator_and_version("< 3")
+    (<built-in function lt>, '3')
+    >>> get_operator_and_version("!= dev")
+    (<built-in function ne>, 'dev')
+    """
+    regexp = r"([<>=!]+)([\w\.]+)"
+    m = re.search(regexp, ver_spec.replace(" ", ""))
+
+    if m is None:
+        raise ValueError(
+            "Invalid value for `ver_spec`: '{}'. Must match this regular expression: '{}'".format(
+                ver_spec, regexp,
+            )
+        )
+
+    return str_to_operator(m.group(1)), m.group(2)
+
+
+def process_requirements(requirements, version=None):
+    """
+    Examples
+    --------
+    >>> process_requirements(None)
+    []
+    >>> process_requirements(["foo"])
+    ['foo']
+    >>> process_requirements({"== 0.1": ["foo"]}, "0.1")
+    ['foo']
+    >>> process_requirements({"< 0.2": ["foo"]}, "0.1")
+    ['foo']
+    >>> process_requirements({"> 0.1, != 0.2": ["foo"]}, "0.3")
+    ['foo']
+    >>> process_requirements({"== 0.1": ["foo"], "== 0.2": ["bar"]}, "0.2")
+    ['bar']
+    >>> process_requirements({"== dev": ["foo"]}, "0.1")
+    []
+    >>> process_requirements({"< dev": ["foo"]}, "0.1")
+    ['foo']
+    >>> process_requirements({"> 0.1": ["foo"]}, "dev")
+    ['foo']
+    >>> process_requirements({"== dev": ["foo"]}, "dev")
+    ['foo']
+    >>> process_requirements({"> 0.1, != dev": ["foo"]}, "dev")
+    []
+    """
+    if requirements is None:
+        return []
+
+    if isinstance(requirements, list):
+        return requirements
+
+    dev_numeric = "9999.9999.9999"
+
+    if version == "dev":
+        version = dev_numeric
+
+    if isinstance(requirements, dict):
+        for ver_spec, packages in requirements.items():
+            op_and_ver_pairs = map(get_operator_and_version, ver_spec.split(","))
+            match_all = all(
+                op(LooseVersion(version), LooseVersion(dev_numeric if ver == "dev" else ver))
```

Comment:
Is it ever the case that `ver == 'dev'`? I don't currently see the use of "dev" as a version in the `requirements` section of any entries in `ml-package-versions.yml`. Do we need this?

Reply:
For example,

```
xgboost:
  ...

  models:
    ...
    requirements:
      "> 1.2.0, != dev": ["XXX"]  # we need XXX in xgboost > 1.2.0 but don't need it in the dev version

```

Reply:
Got it. Is there a case where the requirements for our MLflow tests would be different for the dev version specifically? The library requirements for the dev version are automatically installed when we run the `install_dev` command, so I think this is only relevant to MLflow test requirements.